### ISSUE TITLE
Validate codables

### ISSUE THREAD

**Issue by svanimpe** [IBM-Swift/Kitura#1325]
Most of my usages of codable forms are as follows:

```
guard let form = try? request.read(as: SomeForm.self), form.isValid else {
    response.status(.badRequest)
    return next()
}
```

I was wondering if it makes sense to formalise/generalise this:

- Add a `Validatable` protocol with an `isValid` property.
- When decoding a type that also conforms to `Validatable`, throw an error if the validation fails.

`request.read` already checks if the form's fields are of the correct type, so it makes sense to check the values as well. If conforming to a protocol could get Kitura to do the validation automatically, that would increase safety, as it's easy to forget to add `form.isValid` to every read operation.

Note that I only use this for simple validation checks (that the type can do itself). If validation has dependencies (such as a database connection), I do it separately. I don't think this can be generalised without a proper dependency injection system.

**Comment by ianpartridge** [comment]
Thanks for your idea, it sounds interesting.  I'm wondering why you couldn't implement this solution in your app without it being part of Kitura?  You could define `Validatable` yourself and write a helper that you use inside your route handlers?

**Comment by svanimpe** [comment]
I could and I probably will :) I was just wondering if this is general purpose enough to add to Kitura.

**Comment by ianpartridge** [comment]
OK, let us know how the pattern works in practice. If it's generally useful we could certainly consider adding it to Kitura.

**Comment by svanimpe** [comment]
I gave this a try, and while it does work, I feel it's a bit too cumbersome to do on my end.

I used the following protocol and extension:

```
protocol Validatable {
    var isValid: Bool { get }
}

extension RouterRequest {
    func readAndValidate<T: Decodable & Validatable>(as type: T.Type) throws -> T {
        let result = try read(as: T.self)
        guard result.isValid else {
            throw ServerError.invalidForm
        }
        return result
    }
}
```

I have two issues here:

1. I had to create a new error type here, which doesn't really fit my application. In my application, errors are only used for internal server errors (things that should not happen). User errors are reported with a 400 Bad Request. While I can still do that, it feels silly to have to create and throw an error, just to catch it and return a 400.
2. Ideally, I'd want the new `readAndValidate` method to be an overload of the existing `read` method, where one takes a `Decodable & Validatable` and the other a `Decodable`. This way, I don't have to remember to call the right method, which defeats the purpose of not having to remember to call `isValid`. Unfortunately, this is not something I can do on my end, as I cannot delegate from the former to the latter (the method will just call itself). I could ofcourse copy/paste the implementation of `read` into my `read`, but that's not a clean solution.

**Comment by ianpartridge** [comment]
Thanks for looking into this. I had another thought - as your `SomeForm` is `Decodable`, why not implement the validating logic in a [`init(from:)`](https://developer.apple.com/documentation/swift/decodable/2894081-init) method instead of letting the compiler synthesise it?

**Comment by svanimpe** [comment]
I was hoping to build a simple validation protocol/framework that could be used to validate anything that needs validation.

It would be nice if there was some declarative way to express validation constraints. Java, for example, has a Bean Validation API that lets you add validation annotations to domain types. Instances of these types can then be validated whenever validation is needed, such as when parsing a form of before persisting to a database.

Have you look at Vapor's validation framework (https://docs.vapor.codes/3.0/validation/overview/)? Kitura could benefit from something like this. It would be even better if, in the future, Kitura and Vapor could share a common framework :D

**Comment by ianpartridge** [comment]
@djones6 @Andrew-Lees11 please give your thoughts :-)

**Comment by Andrew-Lees11** [comment]
So this sounds like a good idea. The point of read as and Codable routes is that you get the object you expect and if you have constraints beyond the type it makes sense for you to add that when you define your object. Normally you could add this to the init but because it is codable the init from decoder would be fairly complicated. 

I assume in your implementation isValid would be a computed property where you check the fields of your object. Would this make more sense as a function? 

lastly, I like the idea of overloading read and making a validation framework to make it easier for the user.

**Comment by svanimpe** [comment]
> I assume in your implementation isValid would be a computed property where you check the fields of your object. Would this make more sense as a function?

Yes it was a computed property. I don't think it makes more sense as a function, as that would indicate the implementation can be quite complex (whereas computed properties should be simple in implementation).

Vapor's implementation is interesting too. It's definitely more powerful, but for simple use cases, it feels over engineered, making it hard to use.

**Comment by Andrew-Lees11** [comment]
Ok I'm happy with it as a computed property. I think in most cases you are just going to be checking the value of your fields so having a computed property seems sufficient for this case. Would you like to raise a PR with your suggested change?

**Comment by djones6** [comment]
@ianpartridge A reason to avoid validation during `init(from:)`: the TypeDecoder (used during Swagger generation) works by creating an instance using dummy values.  Validation during init would cause this to fail, or require a mechanism to provide your own (valid) dummy values.

@svanimpe Apologies being late to the party...

I think it makes a lot of sense for this to be a general purpose mechanism, and to have Kitura's Codable routing recognize when types are `Validatable`, and only pass instances to the user's route handler that pass validation.

My preference would be for a mechanism that can communicate back the nature of a validation failure.  A simple (synchronous) implementation using Error might be something like:
```swift
public protocol Validatable {
    func validate() throws
}

struct SomeForm: Codable, Validatable {
  let name: String
  func validate() throws {
    guard !name.isEmpty else {
      throw ValidationError("Name field cannot be empty")
    }
  }
}

do {
  let form = try request.read(as: SomeForm.self)
  try form.validate()
  // continue processing
} catch let error as ValidationError {
  Log.error("Invalid form received: \(error)")
} catch {
  Log.error("Unable to decode form: \(error)")
}
```
I wonder however if it's worth making it asynchronous - what if validation required you to consult some third party? For example, for a registration form, checking an e-mail blacklist service or validating an address against a property or postcode database.  Here's an async version:
```swift
public protocol Validatable {
    func validate(completion: @escaping (ValidationError?) -> Void)
}

struct SomeForm: Codable, Validatable {
    let name: String
    func validate(completion: @escaping (ValidationError?) -> Void) {
        guard !name.isEmpty else {
            return completion(ValidationError("Name cannot be empty"))
        }
        completion(nil)
    }
}

let form = try? request.read(as: SomeForm.self)
form.validate { (error) in
    if let error = error {
        Log.error("Invalid form received: \(error)")
        return next()
    }
    // continue processing
}
```

**Comment by svanimpe** [comment]
Personally, I'd prefer to stick to simple, synchronous, O(1) validations that can be done using only the properties of the form. From a design standpoint, I really wouldn't want a simple form type to require access to the database or some other service. If more complex validation is required, I do it in the route handler that uses the form. That handler already has access to these services.

**Comment by djones6** [comment]
@ianpartridge and I discussed this.  Something we'd like to be able to do is to represent validations in the OpenAPI (swagger) specification that Kitura can generate for a REST API.

When your Codable route handler is registered with the router, the types declared in your handler are inspected by the TypeDecoder and expressed as definitions (models) within the swagger document.  We'd want to also be able to express the constraints on the values for those definitions by deriving those from the code that is used to validate them.

To do this, we'd need an API where the constraints are registered on the Validatable type, and then executed against an instance of that type.  Although I haven't looked closely at Vapor's API, it seems similar in many respects.  However I'd want to keep it simple and accessible.  I'm not sure yet whether those two goals are compatible.
